### PR TITLE
drivers: wdt_tco: Use DT_DRV_COMPAT interface 

### PR DESCRIPTION
**Pull request by finikorg**
DT_DRV_COMPAT is unused for this driver. To get a node identifier DT_NODELABEL() macro is used.

### CODE REVIEW

**Review by martinjaeger — COMMENTED** [review]
Would it not be better to actually get the instance based on the COMPAT instead of the DT_NODELABEL? That would be the more usual Zephyr way, IMO.

**Comment by mbolivar-ampere** [comment]
> > Would it not be better to actually get the instance based on the COMPAT instead of the DT_NODELABEL? That would be the more usual Zephyr way, IMO.
> 
> I think so too. The compat string is what tells us if a driver is compatible with a given DTS node. Since `DT_NODELABEL()` doesn't care about the compat, perhaps it's primarily useful outside of a driver (where you're interested in a driver API rather than a specific implementation of the API).

Not quite; node labels are quite useful when the driver doesn't use DT_INST APIs and needs to get the right nodes anyway. In cases like these you can add some safety rails by ensuring the driver `depends on` some SoC related symbol that ensures an `SOC.dtsi` file should be included in `BOARD.dts` that provides the expected node labels. Various Nordic drivers rely on this pattern.

> 
> That said, I guess we don't have a DT API to ask for a given label but only if it matches the driver's compat? In that case I suppose we'd need to use the instance-specific APIs (`DT_INST()`)? Do @mbolivar-ampere or @galak have any thoughts on this?

I think it's not relevant to this pull request as explained below, but the answer is you can combine `COND_CODE_1` with `DT_NODE_HAS_COMPAT` to do this.
 
> Note: the TCO watchdog driver only supports a single instance, so `DT_NODELABEL(tco_wdt)` would become `DT_INST(0)`.

It sounds like there is no problem with moving to DT_INST APIs, then? In that case I agree that is a better way. (Some HAL-based drivers cannot use DT_INST APIs because they rely on helper functions that need to know the exact instance of a peripheral they are operating on, which is the source of the use of DT_NODELABEL in drivers.)

**Comment by jhedberg** [comment]
@mbolivar-ampere thanks for the input! @finikorg let's go with `DT_INST(0)` then.

> Some HAL-based drivers cannot use DT_INST APIs because they rely on helper functions that need to know the exact instance of a peripheral they are operating on, which is the source of the use of DT_NODELABEL in drivers.)

Some HAL-based Intel drivers have this requirement too, however there it's been solved by adding a `peripheral-id` property: https://github.com/zephyrproject-rtos/zephyr/blob/5e01466d977f04c28fb9f0adf872623f82ae3f51/dts/bindings/i2c/intel%2Csedi-i2c.yaml#L20-L23

File: drivers/watchdog/wdt_tco.c
Comment (jhedberg) [inline drivers/watchdog/wdt_tco.c:270]: I was thinking more in terms of just a one line change: `#define TCO_WDT_NODE DT_INST(0)` - wouldn't that work?

Comment (finikorg) [reply drivers/watchdog/wdt_tco.c:268]: I check current code and have not found this pattern.

Comment (jhedberg) [reply drivers/watchdog/wdt_tco.c:268]: Hardcoding the "magic" `0` number in multiple places doesn't seem ideal to me, whereas having the clear define (like the driver has now) seems like the most intuitive way of communicating to the reader that only a single instance is supported.

Comment (jhedberg) [reply drivers/watchdog/wdt_tco.c:268]: @finikorg can you point me to some of these "single instance supported" drivers that scatter the `0` value across the driver code?

Comment (finikorg) [reply drivers/watchdog/wdt_tco.c:268]: I just searched for "DT_INST_DEFINE(0". However I can use proposed approach with DT_INST(0), I just do not see any pattern of using it.

**Comment by mbolivar-ampere** [comment]
> @mbolivar-ampere thanks for the input! @finikorg let's go with `DT_INST(0)` then.
> 
> > Some HAL-based drivers cannot use DT_INST APIs because they rely on helper functions that need to know the exact instance of a peripheral they are operating on, which is the source of the use of DT_NODELABEL in drivers.)
> 
> Some HAL-based Intel drivers have this requirement too, however there it's been solved by adding a `peripheral-id` property:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/5e01466d977f04c28fb9f0adf872623f82ae3f51/dts/bindings/i2c/intel%2Csedi-i2c.yaml#L20-L23


@jhedberg that's not OK. We do not want DT properties that are there to satisfy HAL implementations. The only reason the nordic drivers use node labels is exactly because this kind of hack was rejected by @galak

**Comment by jhedberg** [comment]
> > Some HAL-based Intel drivers have this requirement too, however there it's been solved by adding a `peripheral-id` property:
> > https://github.com/zephyrproject-rtos/zephyr/blob/5e01466d977f04c28fb9f0adf872623f82ae3f51/dts/bindings/i2c/intel%2Csedi-i2c.yaml#L20-L23
> 
> @jhedberg that's not OK. We do not want DT properties that are there to satisfy HAL implementations. The only reason the nordic drivers use node labels is exactly because this kind of hack was rejected by @galak

@likongintel @kwd-doodling @nashif if the above is the agreed policy, it seems like you will need to remove the `peripheral-id` property from all SEDI drivers and replace it with something else.

**Review by jhedberg — APPROVED** [review]

**Comment by kwd-doodling** [comment]
> > > Some HAL-based Intel drivers have this requirement too, however there it's been solved by adding a `peripheral-id` property:
> > > https://github.com/zephyrproject-rtos/zephyr/blob/5e01466d977f04c28fb9f0adf872623f82ae3f51/dts/bindings/i2c/intel%2Csedi-i2c.yaml#L20-L23
> > 
> > 
> > @jhedberg that's not OK. We do not want DT properties that are there to satisfy HAL implementations. The only reason the nordic drivers use node labels is exactly because this kind of hack was rejected by @galak
> 
> @likongintel @kwd-doodling @nashif if the above is the agreed policy, it seems like you will need to remove the `peripheral-id` property from all SEDI drivers and replace it with something else.

peripheral-id is used to label the real HW ID. From such point, it's HW information.
"the nordic drivers use node labels", which exact driver is it? I'd like to take a look. @mbolivar-ampere

**Comment by mbolivar-ampere** [comment]
> From such point, it's HW information.

It's not considered HW information unless it's a value that needs to be plugged into a register somewhere. If the only reason the property exists is to satisfy an implementation constraint in a software HAL, then it's designing the DT to serve the software instead of the other way around. We generally avoid that if possible, which is why this pattern is not accepted.

> "the nordic drivers use node labels", which exact driver is it? I'd like to take a look

Thanks for taking a look. See e.g. in drivers/spi/spi_nrfx_spim.c

```c

 576   │ /*
 577   │  * We use NODELABEL here because the nrfx API requires us to call
 578   │  * functions which are named according to SoC peripheral instance
 579   │  * being operated on. Since DT_INST() makes no guarantees about that,
 580   │  * it won't work.
 581   │  */
 582   │ #define SPIM(idx)           DT_NODELABEL(spi##idx)
 583   │ #define SPIM_PROP(idx, prop)        DT_PROP(SPIM(idx), prop)
 584   │ #define SPIM_HAS_PROP(idx, prop)    DT_NODE_HAS_PROP(SPIM(idx), prop)
```

**Comment by kwd-doodling** [comment]
> > From such point, it's HW information.
> 
> It's not considered HW information unless it's a value that needs to be plugged into a register somewhere. If the only reason the property exists is to satisfy an implementation constraint in a software HAL, then it's designing the DT to serve the software instead of the other way around. We generally avoid that if possible, which is why this pattern is not accepted.
> 
> > "the nordic drivers use node labels", which exact driver is it? I'd like to take a look
> 
> Thanks for taking a look. See e.g. in drivers/spi/spi_nrfx_spim.c
> 
> ```c
>  576   │ /*
>  577   │  * We use NODELABEL here because the nrfx API requires us to call
>  578   │  * functions which are named according to SoC peripheral instance
>  579   │  * being operated on. Since DT_INST() makes no guarantees about that,
>  580   │  * it won't work.
>  581   │  */
>  582   │ #define SPIM(idx)           DT_NODELABEL(spi##idx)
>  583   │ #define SPIM_PROP(idx, prop)        DT_PROP(SPIM(idx), prop)
>  584   │ #define SPIM_HAS_PROP(idx, prop)    DT_NODE_HAS_PROP(SPIM(idx), prop)
> ```
SoC spec has index assigned to devices, like I2C0/1/2, SPI0/1/2. Could them be treated as HW peripheral-id?
It takes me a while to understand how it work in the spi driver. It seems peripheral-id makes things straight forward.

**Comment by jhedberg** [comment]
> SoC spec has index assigned to devices, like I2C0/1/2, SPI0/1/2. Could them be treated as HW peripheral-id? It takes me a while to understand how it work in the spi driver. It seems peripheral-id makes things straight forward.
> 
> and more, device index is used in PM registers. For example, ISH has a PM register to set PG for I2C devices, each of them is PG controlled by one bit of the register with bit offset same as its device index.

At least to me this sounds like a potentially acceptable reason for having this in DTS, but I'll let @mbolivar-ampere comment. 

Btw, this topic is out-of-scope for this PR, so maybe best moved into a new issue? @martinjaeger could you take another look and approve if you're ok? Thanks.

**Review by martinjaeger — APPROVED** [review]

**Pull request merged**
